src/language/tsLanguageService.ts

```diff
@@ -0,0 +1,184 @@
+const contents: { [name: string]: string } = {};
+
+export default class TypeScriptService {
+  _connection: Connection;
+  _host: TypeScriptServiceHost;
+
+  constructor(connection: Connection) {
+    this._host = this._getTypeScriptServiceHost();
+    this._connection = connection;
+  }
+
+  _loadLibrary(name: string) {
+    let content = contents[name];
+    if (typeof content !== 'string' && librarPath) {
+      const libPath = join(librarPath, name); // From source.
+      try {
+        content = readFileSync(libPath).toString();
+      } catch (e) {
+        this._connection.console.error(
+          `Unable to load library ${name} at ${libPath}`
+        );
+        content = '';
+      }
+      contents[name] = content;
+    }
+    return content;
+  }
+
+  _getTypeScriptServiceHost() {
+    const compilerOptions = {
+      allowNonTsExtensions: true,
+      allowJs: true,
+      lib: ['lib.es2020.full.d.ts'], // Should match to lib from tsconfig.json.
+      target: ts.ScriptTarget.Latest,
+      moduleResolution: ts.ModuleResolutionKind.Classic,
+      experimentalDecorators: false,
+    };
+    let currentTextDocument = TextDocument.create('init', 'javascript', 1, '');
+
+    // Create the language service host to allow the LS to communicate with the host.
+    const host: ts.LanguageServiceHost = {
+      getCompilationSettings: () => compilerOptions,
+      getScriptFileNames: () => [currentTextDocument.uri],
+      getScriptKind: () => {
+        return ts.ScriptKind.JS;
+      },
+      getScriptVersion: (fileName: string) => {
+        if (fileName === currentTextDocument.uri) {
+          return String(currentTextDocument.version);
+        }
+        return '1';
+      },
+      getScriptSnapshot: (fileName: string) => {
+        let text = '';
+        if (fileName === currentTextDocument.uri) {
+          text = currentTextDocument.getText();
+        } else {
+          text = this._loadLibrary(fileName);
+        }
+        return {
+          getText: (start, end) => text.substring(start, end),
+          getLength: () => text.length,
+          getChangeRange: () => undefined,
+        };
+      },
+      getCurrentDirectory: () => '',
+      getDefaultLibFileName: () => 'es2020.full', // Should match to lib from tsconfig.json.
+      readFile: (path: string): string | undefined => {
+        if (path === currentTextDocument.uri) {
+          return currentTextDocument.getText();
+        }
+        return this._loadLibrary(path);
+      },
+      fileExists: (path: string): boolean => {
+        if (path === currentTextDocument.uri) {
+          return true;
+        }
+        return !!this._loadLibrary(path);
+      },
+      directoryExists: (path: string): boolean => {
+        // Typescript tries to first find libraries in node_modules/@types and node_modules/@typescript.
+        if (path.startsWith('node_modules')) {
+          return false;
+        }
+        return true;
```

Comment:
Should this return `true` unconditionally?

This is very interesting though! We might be able to avoid having to implement all the stuff we are adding here in `visitor.ts` if we just manage to provide Typescript with an extra definition file, e.g. something like

```js
/// <reference types="es2020.full" />
declare interface Collection {
  find(query: Record<string, unknown>, project: Record<string, unknown>, options: Record<String, unknown>): any;
  aggregate(pipeline: Record<string, unknown>[], options: Record<String, unknown>): any;
}

declare global {
  let db: {
    [key: string]: Collection;
  };

  let use: (dbName: string) => void;
}
```

this would hopefully make it a lot easier to add definitions for other shell commands as well, so that we don’t have to inspect the AST manually every time

Reply:
Hah I just saw this comment and this is exactly what I left as a comment about above when I added mongodb.d.ts to the lib array, nice 👍